arch/riscv/include/kernel_arch_func.h

```diff
@@ -26,18 +26,27 @@ static ALWAYS_INLINE void arch_kernel_init(void)
 {
 }
 
+#ifdef CONFIG_USE_SWITCH
+extern void z_riscv_switch(void *switch_to, void **switched_from);
+
+static inline void arch_switch(void *switch_to, void **switched_from)
+{
+	z_riscv_switch(switch_to, switched_from);
+}
+#else /* CONFIG_USE_SWITCH */
 static ALWAYS_INLINE void
 arch_thread_return_value_set(struct k_thread *thread, unsigned int value)
 {
 	thread->arch.swap_return_value = value;
 }
+#endif /* CONFIG_USE_SWITCH */
 
 FUNC_NORETURN void z_riscv_fatal_error(unsigned int reason,
 				       const z_arch_esf_t *esf);
 
 static inline bool arch_is_in_isr(void)
 {
-	return _kernel.cpus[0].nested != 0U;
+	return arch_curr_cpu()->nested != 0U;
```

Comment:
In SMP, accessing per-CPU variable (`nested`) should be protected from CPU changing.
There is a race if the thread accesses other CPU's `nested` variable because it is preempted between getting cpu id and accessing `nested` variable.
Zephyr SMP use `arch_irq_lock()` to protect it to be non-preemptive.

Also see [x86 implementation](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.4.0/arch/x86/include/kernel_arch_func.h#L21) or #22565

Reply:
Thanks, indeed. I'll fix it.
Maybe there are other parts that need to be protected from preemption...

Reply:
I find IPI boardcast may need protection.
Also, get CPU ID by `_current_cpu->id` will also do the assertion check of preemption by `z_smp_cpu_mobile()`